### PR TITLE
Support range with sort, limit and count-only

### PR DESCRIPTION
**Pull request by themanforfree**

### CODE REVIEW

File: xline/src/storage/kvstore.rs [hunk 1]
```diff
@@ -237,7 +237,46 @@ impl KvStoreInner {
             count: kvs.len().cast(),
             ..RangeResponse::default()
         };
-        response.kvs = kvs;
+        if !req.count_only {
```
Comment (pzheng1025) [inline xline/src/storage/kvstore.rs:240]: Please check how etcd handle this.

File: xline/src/rpc/mod.rs [hunk 2]
```diff
@@ -71,6 +71,7 @@ pub(crate) use self::etcdserverpb::{
     compare::{CompareResult, CompareTarget, TargetUnion},
     kv_server::{Kv, KvServer},
     lease_server::{Lease, LeaseServer},
+    range_request::{SortOrder, SortTarget},
```
Comment (pzheng1025) [inline xline/src/rpc/mod.rs:74]: only include range_request. And use range_request::SortOrder and range_request::SortTarget.

File: xline/src/storage/kvstore.rs (hunk 1 above)
Comment (themanforfree) [reply xline/src/storage/kvstore.rs:240]: [etcd](https://github.com/etcd-io/etcd/blob/main/server/etcdserver/txn/txn.go#L166-L205)  
here is the code for etcd to handle this. I think it's more complicated than the current implementation

File: xline/src/storage/kvstore.rs [hunk 3]
```diff
@@ -237,7 +237,46 @@ impl KvStoreInner {
...
+            match (req.sort_target(), req.sort_order()) {
```
Comment (pzheng1025) [inline xline/src/storage/kvstore.rs:241]: use req.sort_target/req.sort_order directly?

File: xline/src/storage/kvstore.rs [hunk 4]
```diff
@@ -237,7 +237,46 @@ impl KvStoreInner {
...
+                    kvs.sort_by(|a, b| b.mod_revision.cmp(&a.mod_revision));
+                }
+                (SortTarget::Value, SortOrder::Ascend | SortOrder::None) => {
+                    kvs.sort_by(|a, b| a.value.cmp(&b.value));
+                }
+                (SortTarget::Value, SortOrder::Descend) => {
+                    kvs.sort_by(|a, b| b.value.cmp(&a.value));
+                }
+            }
+            response.kvs = if req.limit > 0 && kvs.len() > req.limit.cast() {
```
Comment (pzheng1025) [inline xline/src/storage/kvstore.rs:274]: req.limit > 0 && kvs.len() > req.limit.cast()   => (req.limit > 0) && (kvs.len() > req.limit.cast()) to make it clearer.

File: xline/src/storage/kvstore.rs [hunk 5]
```diff
@@ -237,7 +237,46 @@ impl KvStoreInner {
...
+                kvs.into_iter().take(req.limit.cast()).collect()
```
Comment (pzheng1025) [inline xline/src/storage/kvstore.rs:275]: kvs.resize() is better? Don't need to clone.

File: xline/src/storage/kvstore.rs (hunk 3 above)
Comment (themanforfree) [reply xline/src/storage/kvstore.rs:241]: Use them directly will match the i32 type, which works but is hard for humans to read.

File: xline/src/storage/kvstore.rs (hunk 5 above)
Comment (themanforfree) [reply xline/src/storage/kvstore.rs]: When limit is less than len, resize will call truncate, so it is better to call truncate directly

**Review by pzheng1025 — APPROVED** [review]

**Pull request merged**